### PR TITLE
[Ready to merge] Pruned-transducer-stateless2 recipe for aidatatang_200zh 

### PR DESCRIPTION
**Pull request by luomingshuang**
This PR is to facilitate code review and it also inherits from the PR #355.  This pr aims to merge into the master. The results and comparisons are as follows. Next, I plan to use a conformer-ctc model for aidatatang_200zh to generate a new baseline.

The WERs are

|                                    |     dev    |    test    | comment                                  |
|------------------------------------|------------|------------|------------------------------------------|
|          greedy search             | 5.53       | 6.59       | --epoch 29, --avg 19, --max-duration 100 |
| modified beam search (beam size 4) | 5.27       | 6.33       | --epoch 29, --avg 19, --max-duration 100 |
| fast beam search (set as default)  | 5.30       | 6.34       | --epoch 29, --avg 19, --max-duration 1500|

The results with [kaldi](https://github.com/kaldi-asr/kaldi/blob/master/egs/aidatatang_200zh/s5/RESULTS): 
```
%WER 37.09 [ 173936 / 468933, 4868 ins, 31143 del, 137925 sub ] exp/mono/decode_test/cer_10_0.0
%WER 17.98 [ 84305 / 468933, 4724 ins, 12637 del, 66944 sub ] exp/tri1/decode_test/cer_13_0.0
%WER 17.94 [ 84149 / 468933, 5025 ins, 12427 del, 66697 sub ] exp/tri2/decode_test/cer_13_0.0
%WER 17.26 [ 80945 / 468933, 4421 ins, 12958 del, 63566 sub ] exp/tri3a/decode_test/cer_14_0.0
%WER 14.16 [ 66424 / 468933, 4567 ins, 10224 del, 51633 sub ] exp/tri4a/decode_test/cer_14_0.0
%WER 12.22 [ 57304 / 468933, 4799 ins, 8197 del, 44308 sub ] exp/tri5a/decode_test/cer_14_0.0
%WER 5.59 [ 26232 / 468933, 1701 ins, 4377 del, 20154 sub ] exp/chain/tdnn_1a_sp/decode_test/cer_10_0.0

# nnet3 tdnn with online pitch, local/nnet3/tuning/run_tdnn_2a.sh
%WER 7.21 [ 33797 / 468933, 2141 ins, 6117 del, 25539 sub ] exp/nnet3/tdnn_sp/decode_test/cer_13_0.0
%WER 7.44 [ 34878 / 468933, 2252 ins, 5854 del, 26772 sub ] exp/nnet3/tdnn_sp_online/decode_test/cer_12_0.0
%WER 7.79 [ 36542 / 468933, 2527 ins, 5674 del, 28341 sub ] exp/nnet3/tdnn_sp_online/decode_test_per_utt/cer_12_0.0

# chain with online pitch, local/chain/tuning/run_tdnn_2a.sh
%WER 5.61 [ 26311 / 468933, 1773 ins, 4789 del, 19749 sub ] exp/chain/tdnn_2a_sp/decode_test/cer_11_0.0
%WER 5.69 [ 26661 / 468933, 1723 ins, 4724 del, 20214 sub ] exp/chain/tdnn_2a_sp_online/decode_test/cer_11_0.0
%WER 5.98 [ 28046 / 468933, 2031 ins, 4527 del, 21488 sub ] exp/chain/tdnn_2a_sp_online/decode_test_per_utt/cer_11_0.0
```

The results with [espnet](https://github.com/espnet/espnet/blob/master/egs/aidatatang_200zh/asr1/RESULTS.md):
Conformer_encoder+SpecAugment + Transformer_decoder
|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|decode_dev_decode_lm|24216|234524|95.8|3.6|0.5|0.1|4.3|21.8|
|decode_test_decode_lm|48144|468933|95.2|4.3|0.5|0.2|5.0|24.0|

### CODE REVIEW

File: egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/asr_datamodule.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+        test = K2SpeechRecognitionDataset(
+            input_strategy=OnTheFlyFeatures(Fbank(FbankConfig(num_mel_bins=80)))
+            if self.args.on_the_fly_feats
+            else PrecomputedFeatures(),
+            return_cuts=self.args.return_cuts,
+        )
+        sampler = DynamicBucketingSampler(
+            cuts,
+            max_duration=self.args.max_duration,
+            rank=0,
```
Comment (csukuangfj) [inline egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/asr_datamodule.py:389]: I think piotr has suggested you to remove rank and world_size here.

File: egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/beam_search.py [hunk 2]
```diff
@@ -0,0 +1,955 @@
+# Copyright    2021  Xiaomi Corp.        (authors: Fangjun Kuang)
```
Comment (csukuangfj) [inline egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/beam_search.py:1]: Could you replace it with a symlink?

**Comment by luomingshuang** [comment]
Thanks, all of requirements are done and tested by running.

**Comment by danpovey** [comment]
If you have a model with attentin_dim=512, it may be too large; you could try 256.

File: egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decoder.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
+# Copyright    2021  Xiaomi Corp.        (authors: Fangjun Kuang)
```
Comment (csukuangfj) [inline egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decoder.py:1]: please replace such files with symlinks.

**Comment by luomingshuang** [comment]
> 

I can have a experiment with dim=256 later.

File: egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decoder.py (hunk 3 above)
Comment (luomingshuang) [reply egs/aidatatang_200zh/ASR/pruned_transducer_stateless2/decoder.py]: Done and tested.

**Comment by luomingshuang** [comment]
When using dim=256 for conformer, the WERs (best performance) are

|                                    |     dev    |    test    | comment                                  |
|------------------------------------|------------|------------|------------------------------------------|
|          greedy search             | 6.16       | 7.36       | --epoch 29, --avg 24, --max-duration 100 |
| modified beam search (beam size 4) | 5.66       | 6.75       | --epoch 29, --avg 24, --max-duration 100 |
| fast beam search (set as default)  | 5.85      | 7.08       | --epoch 29, --avg 24, --max-duration 1500|

The results based on dim=256 are worse than dim=512.

**Comment by csukuangfj** [comment]
> When using dim=256 for conformer, the WERs (best performance) are
> 
> dev	test	comment
> greedy search	6.16	7.36	--epoch 29, --avg 24, --max-duration 100
> modified beam search (beam size 4)	5.66	6.75	--epoch 29, --avg 24, --max-duration 100
> fast beam search (set as default)	5.85	7.08	--epoch 29, --avg 24, --max-duration 1500
> The results based on dim=256 are worse than dim=512.

Did you change other settings other than dim=256, i.e., number of encoder layers, dim feedforward?

**Comment by luomingshuang** [comment]
No, just change encoder model dim=512 to 256.

**Comment by luomingshuang** [comment]
I think this PR can be merged.

**Comment by csukuangfj** [comment]
Thanks!

**Comment by csukuangfj** [comment]
Could you upload a torchscript model to
https://huggingface.co/luomingshuang/icefall_asr_aidatatang-200zh_pruned_transducer_stateless2/tree/main/exp
?

**Comment by luomingshuang** [comment]
Done.
> Could you upload a torchscript model to https://huggingface.co/luomingshuang/icefall_asr_aidatatang-200zh_pruned_transducer_stateless2/tree/main/exp ?

**Comment by csukuangfj** [comment]
Thanks!

I have updated https://huggingface.co/spaces/k2-fsa/automatic-speech-recognition
to include aidatatang_200zh.

**Pull request merged**